### PR TITLE
Config file & CL Overrides - All Major Qs Answered

### PR DESCRIPTION
**Pull request by emmahodcroft**
With the desire to have CL arguments override `config` file came a few questions about exactly how this would work for more complicated stuff. I've been working with James on this in detail, and I believe have a proposal that addresses all the goals:

## What is a Colorby

Outline of what we want:
- Sensible default runs when using (the simpler) CL-only
- A way to control color-by options both in CL and config - with CL overriding config
  - This is difficult with current CL default where everything from `--node-data` plus `--extra-traits` stuff is automatically a colorby
- Always auto-include genotype and date if available - user never has to specify<sup>1</sup>
- Easy way to include output from `augur clades` and `augur seqtraits`. 
  - `augur clades` generates `clade_membership` and `clade_annotation`, both internal names users have never seen, and shouldn't have to learn in order to include
  - `augur seqtraits` generates a highly variable number of traits that are on any particular tree (ex: as DRM in TB, there are 10 drugs where resistance is possible, unknown before running how many are on tree and thus need to be included)
  - Users should never have to write names of traits that _aren't in their metadata_ (ex: `clade_membership`) - they are our internal names and users should not have to learn them!
- No data should go on the tree 'in secret' - if it's not a colorby, geo, or filter, it shouldn't be written out - the user may not realise it's there!

<sup>1</sup> User can put them in config if they want to change `title` (can't change `type`)

### Colorbys and Traits
Some of this (in the code at least) has been further confused because we haven't done much to distinguish colorbys and traits. **TRAITS** are data written onto the nodes of the tree. These include nuc and AA muts, numdate, and more. **COLORBYS** are traits that can be used to colour the tree. Colorbys are always traits but traits are not always colorbys. For the map to work, `country` must be a trait, but it does not have to be a colorby.

To clarify this further, `--traits-to-color` has been replaced with `--color-by-metadata`. More details on how this is different below.

### How we decide what's a Colorby

Things like genotype, author, date are handled separately and always included as **traits** and **colorbys** if present - but these are separate from below code flow:

- Everything passed in via `--node-data` is read in as a **trait**
- Any **traits** that are in `metadata` file are excluded
- What remains will be auto-colorbys (ex: `clade_membership`, DRM data) (more below)
- Any requested geo terms (country, region) are added to **traits**
- Only `--color-by-metadata`<sup>2</sup> entries are **colorbys**, plus any auto-colorbys above
- Any **colorbys** missing from **traits** are added to **traits**
- Only **colorbys** and **traits** are written to the JSON (no data is 'secretly' written out)

<sup>2</sup> Or in `config` if using _only_ `config` - override rules addressed below.

Caveats:
- Special treatment is needed for anything to be a node label - both AA and Clade. This will have to continue to be manually detected.
- Special treatment is needed to auto-set `clade_membership` `title` to 'Clade' so it displays more prettily without users having to learn internal names (see above). This will continue to have to be manually detected.

Going forward:
- This would be easy to expand upon. If you change/predict/modify/add things that are in `metadata`, it needs to be explicitly included (ex: `--metadata-color-by`) to be a colorby. If someone makes a new function that generates new internal names (ex: `coalescent_prob`), they will be automatically included if passed to `--node-data` without users having to know about them.
- For the moment if any of these needed manual adjusting (`title` setting or putting as a node label) this would have to be done manually in `export v2`
- In future we can move towards redoing `augur clades` and `augur seqtraits` so that the output JSON includes all the info `export v2` needs to auto-add node labels and set `title` (if needed), and any new functions can use this format

----
## CL and `config` Override Rules

For anything with both a CL and `config` option, the CL option will override the `config`. The only way to get the `config` option to be valid is to not use the CL argument.

There are 3 exceptions (which do not all except in the same way):
- Display defaults
- Filters
- Colorbys

#### Display defaults
- No way to specify in CL. Defaults are set by `auspice`
- Anything set in `config` is used.

#### Filters
- No way to specify in CL. If no `config`, all colorbys that are categorical or boolean will be filters automatically
- If there is a `config`:
  - If there is no `filters` in `config`, behaves as if CL-only, above
  - If there is `filters` in `config`, then only those listed in this get a filter (must also be in `traits`)
  - If there is `filters` in `config` but it is empty, there are no filters (provides a way to have none)

#### Colorbys
- Colorbys can be set in CL (`--metadata-color-by`) but `title` and `type` cannot by set in CL alone.
- If using CL:
  - If no `config`, all entries in `--metadata-color-by` are given guessed types and variable titles
  - If `config`, _only_ entries in `--metadata-color-by` are used, with `type` and `title` taken from `config` if present (If there are more/fewer entries in `config` they are irrelevant.)
- If using `config`:
  - Only entries in `config` `colorings` will be used (with `type` and `title` used if present, and guessed if not)

In short, unless using `config` alone, `config` can provide `title` and `type` to entries already given in CL, but that's it. 
I can give a chart of examples if anyone wants.

-----
##Remaining questions:
Currently it seems like (?) `author` is always a colorby if present. Is this what we want

'Date' is a bit complicated. `numdate` (the decimal date which includes any estimated dates) is auto-included as a trait and colorby. `date` is the column in the metadata: only on tips, non-decimal, may be ambiguous. Currently if users ask to colorby `date` things get weird. How should we handle this? (Not allow users to pass 'date' in to `--metadata-color-by'?)

### CODE REVIEW

**Comment by trvrb** [comment]
@emmahodcroft --- This is amazingly clear and thoughtful. Really well done. I agree with your entire approach here.

**Comment by rneher** [comment]
very useful, thank you! One thought that just crossed my mind: For json's that `augur` produces, we could include trait info into the `json`. We already do this for the `branch_lengths.json`. This remembers the tree, alignment, and clock model.  if we added `traits:{country:categorical}`, would this help?

**Comment by tsibley** [comment]
Thanks for this detailed writeup! My 2¢ on your outstanding questions.

> Currently it seems like (?) `author` is always a colorby if present. Is this what we want?

Since `author` or `authors` is provided as tip metadata by the user and not generated by an augur command, it seems most consistent with what you've described above to only add a colorby for it if mentioned explicitly in `--metadata-color-by`.

> 'Date' is a bit complicated. `numdate` (the decimal date which includes any estimated dates) is auto-included as a trait and colorby. `date` is the column in the metadata: only on tips, non-decimal, may be ambiguous. Currently if users ask to colorby `date` things get weird. How should we handle this? (Not allow users to pass 'date' in to `--metadata-color-by'?)

Forbidding certain magic values feels strange. How do things get weird when there's a colorby `date`? I made the changes below in zika-tutorial and viewed the result in Auspice. I'm not sure it's useful, but it seemed reasonable enough. Are there cases where users may not have `numdate` but only some other `date` in metadata?

```diff
diff --git a/config/auspice_config.json b/config/auspice_config.json
index fbc3f8d..3ae1ef2 100644
--- a/config/auspice_config.json
+++ b/config/auspice_config.json
@@ -8,11 +8,17 @@
       "key": "genotype"
     },
     "num_date": {
-      "menuItem": "date",
-      "legendTitle": "Sampling date",
+      "menuItem": "resolved date",
+      "legendTitle": "Resolved sampling date",
       "type": "continuous",
       "key": "num_date"
     },
+    "date": {
+      "menuItem": "original date",
+      "legendTitle": "Sampling date",
+      "type": "discrete",
+      "key": "date"
+    },
     "authors": {
       "key":"authors",
       "legendTitle":"Authors",
```

**Comment by jameshadfield** [comment]
> Forbidding certain magic values feels strange. 

We have lots of magics -- some of which are ignored, some handled as special cases, some probably not handled yet -- we just need a good list of them & what happens. While some may work if you define as a color-by, they should be discouraged. These are the ones I can think of off the top of my head:

* `paper_url`, `author(s)`, `journal`
* `date`, `num_date`, `numdate`
* `div`, `branch_length`, `clock_length`
* `url`
* `hidden` 
* `clade_annotation`, `clade_membership`, `label(s)`

**Comment by tsibley** [comment]
> we just need a good list of them, what happens, and what you can/can't use as a color-by

Yes, documenting these would be good. Your list is a good start. My position is that aiming to reduce the number of magic values and special fields will make overall usage easier and more consistent.

**Comment by emmahodcroft** [comment]
> very useful, thank you! One thought that just crossed my mind: For json's that augur produces, we could include trait info into the json. We already do this for the branch_lengths.json. This remembers the tree, alignment, and clock model. if we added traits:{country:categorical}, would this help?

Yes, this could be useful to include going forward. Categorical is the default guess so mostly I imagine we will capture these well, but there will be cases where we get this wrong and it would be good to address that. This could be bundled into a future plan of making other augur functions (`sequence-traits`, `clades`, etc) export JSONS that say what `title` should be and if something is a branch label. I think all this is a good direction to aim for, as it means we can just give guidance on the JSON format for any new external or internal function and `augur export` will be able to handle it perfectly.

**Comment by emmahodcroft** [comment]
> Since author or authors is provided as tip metadata by the user and not generated by an augur command, it seems most consistent with what you've described above to only add a colorby for it if mentioned explicitly in --metadata-color-by.

I agree with this - I'll push a commit changing this. 

> Forbidding certain magic values feels strange. How do things get weird when there's a colorby date?

Yes, it's not crazy or breaking behaviour, which is why I settled on "weird". If you include `date` in `--color-by-metadata` you get a colorby option "date" (as well as the usual "Sampling date"), which has the `date` (a string, from `refine`) on internal nodes and `date` (from metadata) on tips and processes them as categorical trait: 
image

My main concern might be that if this comes up before "Sampling date" in the drop-down (for me it does) and "Sampling date" is below many other traits, possibly even scrolled-off, users may think this is the date as it comes out of TimeTree and that something's gone wrong. 

Perhaps the easiest solution to this would be to order the colorby drop-down options so that Genotype and Sample Date always sit at the top. I'm not sure how they are ordered at the moment.

**Comment by emmahodcroft** [comment]
I've added a check so that authors is included as a colorby if requested (not automatically).

Internally we quietly standardise to "author" instead of "authors" some places, so in this same vein I quietly standardise colorby and filter requests for "authors" to be processed as "author". (If users have an 'authors' column it will be interpreted everywhere else as 'author' so it makes sense to also correct their colorby request, since they don't know this is happening...)

---
This did remind me that there is a half-violation of something I said above - `url`, `paper_url`, `accession`, `author`, and `title` are always exported if present regardless of status as colorby, filter, or geo. They are, however, clearly visible when you click on a tip, so I'd say this also falls on the right side of not 'secretly' exporting un-asked-for data.

**Comment by tsibley** [comment]
> My main concern might be that if this comes up before "Sampling date" in the drop-down (for me it does) and "Sampling date" is below many other traits, possibly even scrolled-off, users may think this is the date as it comes out of TimeTree and that something's gone wrong.
> 
> Perhaps the easiest solution to this would be to order the colorby drop-down options so that Genotype and Sample Date always sit at the top. I'm not sure how they are ordered at the moment.

Yeah, it can definitely get confusing. I'm +1 for not forbidding it, but making it less confusing through sorting the color bys in Augur (or Auspice?). Good idea.
